src/util/bson.hpp

```diff
@@ -0,0 +1,1218 @@
+    > {
+        size_t m_idx = 0;
+        IndexedMap* m_map;
+    public:
+        iterator(IndexedMap* map, size_t idx) : m_idx(idx), m_map(map) {}
+        iterator& operator++()
+        {
+            m_idx++;
+            return *this;
+        }
+
+        iterator& operator--()
+        {
+            m_idx--;
+            return *this;
+        }
+
+        iterator operator++(int)
+        {
+            return ++(*this);
+        }
+
+        iterator operator--(int)
+        {
+            return --(*this);
+        }
+
+        bool operator==(iterator other) const
+        {
+            return m_idx == other.m_idx;
+        }
+
+        bool operator!=(iterator other) const
+        {
+            return !(*this == other);
+        }
+
+        std::pair<std::string, T> operator*()
+        {
+            return m_map->operator[](m_idx);
+        }
+    };
+
+    using entry = std::pair<std::string, T>;
+
+    IndexedMap() {}
+    IndexedMap(std::initializer_list<entry> entries) {
+        for (auto entry : entries) {
+            m_keys.push_back(entry.first);
+            m_map[entry.first] = entry.second;
+        }
+    }
+
+    /// The size of the map
+    size_t size() const
+    {
+        return m_map.size();
+    }
+
+    /// Find an entry by index
+    entry operator[](size_t idx)
+    {
+        auto key = m_keys[idx];
+        return { key, m_map[key] };
+    }
+
+    iterator begin()
+    {
+        return iterator(this, 0);
+    }
+
+    iterator end()
+    {
+        return iterator(this, m_map.size());
+    }
+
+    iterator front()
+    {
+        return this->begin();
+    }
+
+    iterator back()
+    {
+        return this->end()--;
+    }
+
+    /// Find or add a given key
+    T& operator[](const std::string& k)
+    {
+        auto entry = m_map.find(k);
+        if (entry == m_map.end()) {
+            m_keys.push_back(k);
+        }
+
+        return m_map[k];
+    }
+
+    /// Whether or not this map is empty
+    bool empty()
+    {
+        return m_map.empty();
+    }
+
+    /// Pop the last entry of the map
+    void pop_back()
+    {
+        auto last_key = m_keys.back();
+        m_keys.pop_back();
+        m_map.erase(last_key);
+    }
+
+    std::vector<std::string> keys() const
+    {
+        return m_keys;
+    }
+private:
+    friend inline bool operator==(const IndexedMap<T>& lhs, const IndexedMap<T>& rhs);
+    std::unordered_map<std::string, T> m_map;
+    std::vector<std::string> m_keys;
+};
+
+template <typename T>
+inline bool operator==(const IndexedMap<T>& lhs, const IndexedMap<T>& rhs)
+{
+    return lhs.m_set == rhs.m_set && lhs.m_keys == rhs.m_keys;
+}
+
+/**
+ ====== BSON =======
+ */
+
+class Bson;
+
+/// A variant of the allowed Bson types.
+class Bson : public std::variant<
+    bson_null_t,
+    int32_t,
+    int64_t,
+    bool,
+    float,
+    double,
+    time_t,
+    const char*,
+    BinaryData,
```

Comment:
Bson needs to own the data for string and binary values, so these need to be `std::string` and `std::vector<char>`.